i18n(es): add sharing-state.md (#1542)
withastro/docs: src/pages/es/core-concepts/sharing-state.md

---
layout: ~/layouts/MainLayout.astro
title: Compartir Estado
i18nReady: true
setup: |
  import Tabs from '../../../components/tabs/Tabs'
  import UIFrameworkTabs from '~/components/tabs/UIFrameworkTabs.astro'
  import LoopingVideo from '~/components/LoopingVideo.astro'
---

> Al construir tu proyecto con la [islands architecture / hidratación parcial](/es/concepts/islands/), puede que te hayas topado con este problema: **Quiero compartir estado entre mis componentes.**

Reviewer: What do you think about replace "islands architecture" with "arquitectura de islas"?

Reviewer 2: he visto que hemos traducido esto como: arquitectura de la isla, arquitectura de islas, y algunos lo hemos dejando en ingles. Seria bueno consolidar en una opcion.

Author: Oh yeah, I thought we were using the term without being translated but we aren't 😓 I'll change it!

Author: Arquitectura de islas como propone @diegognt me parece bien, qué les parece a los demás?

> Al construir tu proyecto con la [islands architecture / hidratación parcial](/es/concepts/islands/), puede que te hayas topado con este problema: **Quiero compartir estado entre mis componentes.**

Reviewer 2: ```suggestion
Al construir tu proyecto usando la [arquitectura de islas / hidratación parcial](/es/concepts/islands/), puede que te hayas topado con este problema: **Quiero compartir estado entre mis componentes.**
```

Astro recomienda una solución diferente para el almacenamiento compartido en el lado del cliente: [**Nano Stores**](https://github.com/nanostores/nanostores).

## ¿Por qué Nano Stores?

La librería [Nano Stores](https://github.com/nanostores/nanostores) te permite crear _stores_ con las que cualquier componente puede interactuar. Recomendamos Nano Stores porque:
- **Son livianas.** Nano Stores tiene la cantidad mínima de JS que puedas llegar a requerir (menos de 1 KB) con cero dependencias.
- **Son framework-agnósticas.** ¡Esto significa que puedes compartir estado entre Preact, Svelte y Vue sin contratiempos! Astro está construido para ser flexible, así que amamos las soluciones que ofrecen una experiencia de desarrollo similar, sin importar tu preferencia.

Aun así, hay otras alternativas a explorar. Entre ellas puedes encontrar:
- [Svelte stores incorporadas](https://svelte.dev/tutorial/writable-stores)
- [Solid signals](https://www.solidjs.com/docs/latest) fuera de un elemento de contexto

> - [Enviar eventos de navegador personalizados](https://developer.mozilla.org/es/docs/Web/Events/Creating_and_triggering_events) entre componentes

Reviewer 2: ```suggestion
- [Enviar eventos del navegador personalizados](https://developer.mozilla.org/es/docs/Web/Events/Creating_and_triggering_events) entre componentes
```

:::note[FAQ]

<details>
<summary>**🙋 ¿Puedo usar Nano Stores en archivos `.astro` u otros archivos del lado del servidor?**</summary>

Las Nano Stores _pueden_ ser importadas, escritas y leídas desde componentes del lado del servidor, **¡aunque no lo recomendamos!**. Esto se debe a ciertas restricciones:

> - Escribir en una store desde un archivo `.astro` o un [componente no-hidratado](/es/core-concepts/framework-components/#hidratando-componentes-interactivos) _no_ afectará el valor recibido por un [componente del lado del cliente](/es/reference/directives-reference/#client-directives).

Reviewer 3: Looks like we need to fix this link:

```suggestion
- Escribir en una store desde un archivo `.astro` o un [componente no-hidratado](/es/core-concepts/framework-components/#hidratando-componentes-interactivos) _no_ afectará el valor recibido por un [componente del lado del cliente](/es/reference/directives-reference/#directivas-del-cliente).
```

¡Si entiendes estas restricciones y aun así encuentras un caso de uso, puedes darle una oportunidad a Nano Stores! Solamente recuerda que Nano Stores fue creado específicamente para reaccionar a cambios en el **cliente**.

</details>

<details>
<summary>**🙋 ¿Cómo se comparan las Svelte stores a Nano Stores?**</summary>

**¡Nano Stores y [Svelte stores](https://svelte.dev/tutorial/writable-stores) son muy similares!** De hecho, [nanostores te permite usar el mismo atajo `$`](https://github.com/nanostores/nanostores#svelte) para suscripciones que puedes utilizar con las Svelte stores.

Si quieres evitar usar una librería de terceros, [Svelte stores](https://svelte.dev/tutorial/writable-stores) es una gran herramienta para la comunicación entre islas. Aun así, puedes llegar a preferir Nano Stores si a) te gustan sus add-ons para ["objetos"](https://github.com/nanostores/nanostores#maps) y [estado asíncrono](https://github.com/nanostores/nanostores#lazy-stores), o b) quieres comunicarte entre Svelte y otros frameworks como Preact o Vue.
</details>

<details>
<summary>**🙋 ¿Cómo se comparan las Solid signals a Nano Stores?**</summary>

> Si has usado Solid un buen rato, habrás intentado mover [signals](https://www.solidjs.com/docs/latest#createsignal) o [stores](https://www.solidjs.com/docs/latest#createstore) fuera de tus componentes. ¡Esta es una muy buena manera de compartir estado entre islas de Solid! Intenta exportar signals desde un archivo compartido:

Reviewer 2: ```suggestion
Si has usado Solid un anteriormente, habrás intentado mover [signals](https://www.solidjs.com/docs/latest#createsignal) o [stores](https://www.solidjs.com/docs/latest#createstore) fuera de tus componentes. ¡Esta es una muy buena manera de compartir estado entre islas de Solid! Intenta exportar signals desde un archivo compartido:
```

```js
// sharedStore.js
import { createSignal } from 'solid-js';

export const sharedCount = createSignal(0);
```

> ...y todos los componentes que importen `sharedCount` compartirán el mismo estado. Aunque esto funciones bien, puedes llegar a preferir Nano Stores si a) te gustan sus add-ons para ["objetos"](https://github.com/nanostores/nanostores#maps) y [estado asíncrono](https://github.com/nanostores/nanostores#lazy-stores), o b) quieres comunicarte entre Solid y otros frameworks como Preact or Vue.

Reviewer 2: ```suggestion
...y todos los componentes que importen `sharedCount` compartirán el mismo estado. Aunque esto funciones bien, puedes llegar a preferir Nano Stores si a) te gustaría añadir add-ons para ["objetos"](https://github.com/nanostores/nanostores#maps) y [estado asíncrono](https://github.com/nanostores/nanostores#lazy-stores), o b) quieres comunicarte entre Solid y otros frameworks como Preact or Vue.
```

<LoopingVideo sources={[{ src: '/videos/stores-example.mp4', type: 'video/mp4' }]} />

_[**Prueba el ejemplo terminado**](https://github.com/withastro/astro/tree/main/examples/with-nanostores) en tu máquina u online vía Stackblitz._

Tu archivo base de Astro podría verse así:

```astro
---
// Example: src/pages/index.astro
import CartFlyoutToggle from '../components/CartFlyoutToggle';
import CartFlyout from '../components/CartFlyout';
import AddToCartForm from '../components/AddToCartForm';
---

<!DOCTYPE html>
<html lang="en">
<head>...</head>
<body>
  <header>
    <nav>
      <a href="/">Tienda de Astro</a>
      <CartFlyoutToggle client:load />
    </nav>
  </header>
  <main>
    <AddToCartForm client:load>
    <!-- ... -->
    </AddToCartForm>
  </main>
  <CartFlyout client:load />
</body>
</html>
```

### Usando "atoms"

Empecemos por abrir nuestro `CartFlyout` cada vez que cliqueamos en `CartFlyoutToggle`. 

Primero, crea un nuevo archivo JS o TS para nuestra store. Usaremos un ["atom"](https://github.com/nanostores/nanostores#atoms) para esto:

```js
// src/cartStore.js
import { atom } from 'nanostores';

export const isCartOpen = atom(false);
```

> Ahora, podemos importar esta store dentro de cualquier archivo que necesite leer o escribir. Comenzaremos conectando nuestro `CartFlyoutToggle`:

Reviewer 2: ```suggestion
Ahora, podemos importar esta store dentro de cualquier archivo que necesite leerlo o sobreescribirlo. Comenzaremos conectando nuestro `CartFlyoutToggle`:
```

Author: Qué te parece `leer o escribir en ella`? Ya que estamos hablando de "la" store (o al menos yo lo interpreté con pronombre femenino)

  }

  return (
    <form onSubmit={addToCart}>
      {children}
    </form>
  )
}
```
</Fragment>
<Fragment slot="svelte">
```svelte
<!--src/components/AddToCartForm.svelte-->
<form on:submit|preventDefault={addToCart}>
  <slot></slot>
</form>

<script>
  import { addCartItem, isCartOpen } from '../cartStore';

  // ¡usaremos valores fijos por simplicidad!
  const hardcodedItemInfo = {
    id: 'astronaut-figurine',
    name: 'Astronaut Figurine',
    imageSrc: '/images/astronaut-figurine.png',
  }

  function addToCart() {
    isCartOpen.set(true);
    addCartItem(hardcodedItemInfo);
  }
</script>
```
</Fragment>
<Fragment slot="vue">
```vue
<!--src/components/AddToCartForm.vue-->
<template>
  <form @submit="addToCart">
    <slot></slot>
  </form>
</template>

<script setup>
  import { addCartItem, isCartOpen } from '../cartStore';

  // ¡usaremos valores fijos por simplicidad!
  const hardcodedItemInfo = {
    id: 'astronaut-figurine',
    name: 'Astronaut Figurine',
    imageSrc: '/images/astronaut-figurine.png',
  }

  function addToCart(e) {
    e.preventDefault();
    isCartOpen.set(true);
    addCartItem(hardcodedItemInfo);
  }
</script>
```
</Fragment>
</UIFrameworkTabs>

> Finalmente, renderizamos los ítems dentro del carrito en nuestro componente `CartFlyout`:

Reviewer 2: ```suggestion
Finalmente, renderizamos los artículos dentro del carrito en nuestro componente `CartFlyout`:
```

Author: Creo que podríamos dejar ítem ya que en los ejemplos de código hay mucho `item`, qué te parece?

Reviewer 2: no creo que haya ningun problema
